Source/WebCore/page/DragController.cpp

```diff
@@ -380,21 +380,23 @@ static bool isInShadowTreeOfEnabledColorInput(Node& node)
 #endif
 
 // This can return null if an empty document is loaded.
-static Element* elementUnderMouse(Document* documentUnderMouse, const IntPoint& p)
+static RefPtr<Element> elementUnderMouse(Document& documentUnderMouse, const IntPoint& p)
 {
-    auto* frame = documentUnderMouse->frame();
+    auto* frame = documentUnderMouse.frame();
     float zoomFactor = frame ? frame->pageZoomFactor() : 1;
     LayoutPoint point(p.x() * zoomFactor, p.y() * zoomFactor);
 
     HitTestResult result(point);
-    documentUnderMouse->hitTest(HitTestRequest(), result);
+    documentUnderMouse.hitTest(HitTestRequest(), result);
 
     auto* node = result.innerNode();
     if (!node)
         return nullptr;
     // FIXME: Use parentElementInComposedTree here.
-    auto* element = is<Element>(*node) ? &downcast<Element>(*node) : node->parentElement();
-    auto* host = element->shadowHost();
+    RefPtr element = dynamicDowncast<Element>(*node);
```

Comment:
So much ref'ing / deref'ing but not strictly necessary. Here we only call trivial getters on `element` so I don't think it needs to be a RefPtr.